File: Entities/DiscountRule.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Entities
{
    public class DiscountRule
    {
        private readonly Func<string, bool> condition;
        private readonly decimal discount;

        public DiscountRule(Func<string, bool> condition, decimal discount)
        {
            this.condition = condition;
            this.discount = discount;
        }

        public decimal applyDiscount(string name, decimal cost)
        {

            if (this.condition(name))
            {
                return cost * (1 - this.discount);
            }
            else
            {
                return cost;
            }
        }
    }
}

File: EmployeeCostCalculator/ClientApp/src/app/models/discount-rule.ts
export class DiscountRule{
  //TODO: better use Person or even employee to apply possible future complex rule
  constructor(
    private readonly rule : (name: string) => boolean,
    private readonly discount: number) {
  }

  public applyDiscount(name : string, cost:number) : number {
    if(!cost) return cost;
    if(this.rule(name)) {
      return cost * (1 - this.discount);
    } else {
      return cost;
    }
  }
}
